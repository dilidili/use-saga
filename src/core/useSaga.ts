import { useReducer, Reducer, useEffect, useRef } from 'react';
import { runSaga, stdChannel, Saga } from 'redux-saga';
import { takeEvery, call, put } from 'redux-saga/effects'

interface Action {
  type: string;
  [key: string]: any;
};

type Model<S> = {
  state: any,
  reducers?: {
    [key: string]: Reducer<S, Action>;
  },
  effects?: {
    [key: string]: Saga;
  },
}


export default function useSaga<S = {}>(model: Model<S>): [S, (action: Action) => void] {
  const {
    state: initialState,
    reducers = {},
    effects = {},
  } = model;

  // reducers
  const reducer = (prevState: S, action: Action) => Object.keys(reducers).reduce((r, v) => v === action.type ? reducers[v](prevState, action) : r, prevState);
  const [state, dispatch] = useReducer(reducer, initialState);

  // effects
  const refState = useRef<S>(state);
  const refDispatch = useRef(dispatch);
  const refChannel = useRef(stdChannel());

  useEffect(() => {
    refState.current = state; 
    refDispatch.current = dispatch; 
  }, [state, dispatch]);

  useEffect(() => {
    const boundRunSaga = runSaga.bind(null, {
      channel: refChannel.current,
      dispatch: (...args) => refDispatch.current.apply(null, args),
      getState: () => refState.current,
    });

    boundRunSaga(function *() {
      const keys = Object.keys(effects);
      for(let i = 0; i < keys.length; i++) {
        yield takeEvery(keys[i], function* (action) {
          yield effects[keys[i]](action, { call, put });
        });
      }
    });
  }, []);

  return [state, (action: Action) => {
    refChannel.current.put(action)
  }];
};