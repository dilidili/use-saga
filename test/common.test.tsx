import useSaga, { useFetch, } from '../src';
import React from 'react';
import renderer from 'react-test-renderer';

const Api = {
  fetchUser: function* (id = 'test') {
    return { id };
  },
};

export function* fetchData(action, { call, put }) {
  const data = yield call(Api.fetchUser, action.url);
  yield put({ type: "FETCH_SUCCEEDED", data });
}

export function* fetchDataFail(action, { put }) {
  yield put({ type: "FETCH_FAILED", error: new Error() });
}

test('common AJAX request', function() {
  const Test: React.FC<{ fetchData }> = ({ fetchData }) => {
    const [state, dispatch] = useSaga<{
      data: {
        id: string,
      } | null, 
      error: Error | null,
    }>({
      state: {
        data: null,
        error: null,
      },
      reducers: {
        ['FETCH_SUCCEEDED']: (state, { data }) => {
          return { ...state, data };
        },
        ['FETCH_FAILED']: (state, { error }) => {
          return { ...state, error };
        },
      },
      effects: {
        ['FETCH_REQUESTED']: fetchData,
      },
    });

    return <div onClick={() => dispatch({ type: 'FETCH_REQUESTED', url: 'test' })}>{state.error ? 'error' : state.data ? state.data.id : 'loading'}</div>;
  };

  // success flow
  let component
  renderer.act(() => {
    component = renderer.create(<Test fetchData={fetchData} />);
  });

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  renderer.act(() => {
    tree.props.onClick();
  });

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // fail flow
  let failComponent;
  renderer.act(() => {
    failComponent = renderer.create(<Test fetchData={fetchDataFail} />);
  });

  let failTree = failComponent.toJSON();
  renderer.act(() => {
    failTree.props.onClick();
  });

  failTree = failComponent.toJSON();
  expect(failTree).toMatchSnapshot();
});

test('useFetch helper', () => {
  const Test: React.FC<{ fetchData }> = ({ fetchData }) => {
    const [fetchUser, { start }] = useFetch(fetchData);

    return <div onClick={() => start('leisure life')}>
      <div>{fetchUser.error ? 'error' : fetchUser.data ? fetchUser.data.id : 'empty'}</div>
      <div>{fetchUser.loading ? 'loading' : 'complete'}</div>
    </div>
  };

  // success flow
  let component
  renderer.act(() => {
    component = renderer.create(<Test fetchData={Api.fetchUser} />);
  });

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  renderer.act(() => {
    tree.props.onClick();
  });

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // fail flow
  let failComponent;
  renderer.act(() => {
    failComponent = renderer.create(<Test fetchData={function *() { throw Error() }} />);
  });

  let failTree = failComponent.toJSON();
  renderer.act(() => {
    failTree.props.onClick();
  });

  failTree = failComponent.toJSON();
  expect(failTree).toMatchSnapshot();
});

test('default state updater', () => {
  const Test: React.FC = () => {
    const [{ count }, dispatch] = useSaga({
      state: {
        count: 0,
      },
    });

    return <div onClick={() => dispatch({ type: 'updateCount', count: 1 })}>
      <div>{count}</div>
    </div>
  };

  let component
  renderer.act(() => {
    component = renderer.create(<Test />);
  });
  let tree = component.toJSON();
  renderer.act(() => {
    tree.props.onClick();
  });
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('setState effect', () => {
  const Test: React.FC = () => {
    const [{ count, touched }, dispatch] = useSaga({
      state: {
        count: 0,
        touched: false,
      },

      effects: {
        updateCount: function *({ count }, { setState }) {
          yield setState({
            count,
            touched: true,
          });
        },
      },
    });

    return <div onClick={() => dispatch({ type: 'updateCount', count: 2 })}>
      <div>{count}</div>
      <div>{touched + ''}</div>
    </div>
  };

  let component
  renderer.act(() => {
    component = renderer.create(<Test />);
  });
  let tree = component.toJSON();
  renderer.act(() => {
    tree.props.onClick();
  });
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});