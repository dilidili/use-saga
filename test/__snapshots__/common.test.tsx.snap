// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common AJAX request 1`] = `
<div
  onClick={[Function]}
>
  loading
</div>
`;

exports[`common AJAX request 2`] = `
<div
  onClick={[Function]}
>
  test
</div>
`;

exports[`common AJAX request 3`] = `
<div
  onClick={[Function]}
>
  error
</div>
`;

exports[`default state updater 1`] = `
<div
  onClick={[Function]}
>
  <div>
    1
  </div>
</div>
`;

exports[`setState effect 1`] = `
<div
  onClick={[Function]}
>
  <div>
    2
  </div>
  <div>
    true
  </div>
</div>
`;
